File: Python/4Sum.py
class Solution:
    # @return a list of lists of length 4, [[val1,val2,val3,val4]]
    def fourSum(self, num, target):
        n = len(num)
        num.sort()
        res = []
        for i in range(n-3):
            if i > 0 and num[i] == num[i-1]:
                continue
            for j in range(i+1, n-2):
                if j > i + 1 and num[j] == num[j-1]:
                    continue
                x = j + 1
                y = n - 1
                while x < y:
                    s = num[i] + num[j] + num[x] + num[y]
                    if s == target:
                        res.append([num[i], num[j], num[x], num[y]])
                        x += 1
                        y -= 1
                        while x < y and num[x] == num[x-1]:
                            x += 1
                        while x < y and num[y] == num[y+1]:
                            y -= 1
                    elif s < target:
                        x += 1
                    else:
                        y -= 1
        return res

File: C++/4Sum.cpp
class Solution {
public:
    vector<vector<int> > fourSum(vector<int>& nums, int target) {
        vector<vector<int> > res;
        sort(nums.begin(), nums.end());
        for (int i = 0; i + 3 < nums.size(); i ++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j + 2 < nums.size(); j ++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                int left = j + 1, right = nums.size() - 1;
                while (left < right) {
                    int temp = nums[i] + nums[j] + nums[left] + nums[right];
                    if (temp == target) {
                        res.push_back(vector<int> {nums[i], nums[j], nums[left], nums[right]});
                        int temp1 = nums[left++], temp2 = nums[right--];
                        while (left < right && nums[left] == temp1) left++;
                        while (left < right && nums[right] == temp2) right--;
                    } else if (temp < target) {
                        left ++;
                    } else {
                        right --;
                    }
                }
            }
        }
        return res;
    }
};